docs/exceptions.md

```diff
@@ -50,6 +53,11 @@ It is always a tradeoff whether to allow programmer mistakes, or do type-level
 trickery to avoid them. The decision process for this is out of scope of this
 document.
 
+*Comment:* calling `zipSameLen` with lists of different length is a mistake
```

Comment:
My point is on lines 57/58: this has nothing to do with exception handling (or exceptions in general).

Reply:
According to your definition of exception.

Reply:
> According to your definition of exception.

Yes. It's more faithful to reality.

Reply:
> Yes. It's more faithful to reality.

To your perception of reality. This is leading us nowhere, my understanding is that "exception" is a heavily overloaded term. You find it used with one meaning (an exceptional situation) and then insist that the meaning is different (a term in Haskell with an `Exception` instance, I guess?). Let's be constructive here. How should this document be named, if we keep the current contents?

* exception handling guidelines
* error handling guidelines
* ...

The goal of the document is to help the programmer decide how to structure his/her code when he/she needs to do something vaguely related to bugs/exceptions/errors/throwing/catching/etc. If we narrow it down to dealing with IO exceptions only, for example, it will no longer serve its purpose.

Reply:
> To your perception of reality.

I meant objective reality.

> my understanding is that "exception" is a heavily overloaded term.

It is.

> The purpose of the document is to help the programmer decide how to structure his/her code when he/she needs to do something vaguely related to bugs/exceptions/errors/throwing/catching/etc. If we narrow it down to dealing with IO exceptions only, for example, it will no longer serve its purpose.

`IO` exceptions are the real deal, though. They are substantively different from pure "exceptions":

```Haskell
throwIO e >> pure () = _|_

(throwError e :: ExceptT e Identity x) >> pure () /= _|_
```

If you're writing a pure function, you don't throw exceptions. You ought to be allowed to use `ExceptT`, because "throwing an exception" in this monad is actually pure, despite the name. No user of your pure function will try to catch exceptions when using it, because it's a pure function. It doesn't throw exceptions, it just is.

When a pure partial function like `zipSameLen` blows up due to `error`, it's not an exception, it's a bug. The programmer misused it, but could have avoided doing so. Contrast with `readFile`, in which the programmer could not possibly rule out disk errors, file permission issues, etc.

Reply:
> `throwIO e >> pure () = _|_`

Well, no. 

```
Prelude Control.Exception> (throwIO (ErrorCall "err") >> pure ()) `seq` pure ()
```

`throwIO` is a non-bottom operation.